src/depackers/vorbis.c

```diff
@@ -3655,9 +3645,8 @@ static int start_decoder(vorb *f)
    if (!vorbis_validate(header))                    return error(f, VORBIS_invalid_first_page);
    // vorbis_version
    if (get32(f) != 0)                               return error(f, VORBIS_invalid_first_page);
-   val = get8(f); if (val == 0)                     return error(f, VORBIS_invalid_first_page);
-   if (val > STB_VORBIS_MAX_CHANNELS)               return error(f, VORBIS_too_many_channels);
-   f->channels = val;
+   f->channels = get8(f); if (!f->channels)         return error(f, VORBIS_invalid_first_page);
+   if (f->channels > STB_VORBIS_MAX_CHANNELS)       return error(f, VORBIS_too_many_channels);
```

Comment:
This one is interesting because at first glance it looks like it does *absolutely nothing of value*, but actually did fix invalid free bugs in the old stb_vorbis.

1) `start_decoder` is only called from `stb_vorbis_open_memory` (line 5199) on a stack-allocated `stb_vorbis` that is zero initialized by `vorbis_init`.
2) `start_decoder` returns 0 (line 3658-9) due to a zero or too large channel count. Assume that the channels count is written directly to `f->channels` like upstream does it.
3) Quick check: `error`, defined at line 935, does in fact always return 0.
4) On failure, `stb_vorbis_open_memory` sets `*error`, calls `vorbis_deinit`, and returns NULL.
5) In `vorbis_deinit` at line 4306, `p->channels` is iterated on. This is combined with a check to prevent `i` from ever being >=`STB_VORBIS_MAX_CHANNELS` so no frees are performed on out-of-bounds array positions. The reason for the libxmp hack is that this second check was missing in the old stb_vorbis.
6) Since there's an upstream safety check here, the old check can be removed.

Bonus: `stb_vorbis_open_pushdata` and `stb_vorbis_open_file_section`, the other places that use `start_decoder` (deadcode in our copy), also behave in similar ways, so the only thing using `f->channels` for this error state is that loop in `vorbis_deinit`.

This check was added here: https://github.com/libxmp/libxmp/commit/957e4a61d6f3798937dbca69e2e3716b111563a8.